decisions/C-ManagerInterface.md

```diff
@@ -0,0 +1,205 @@
+# Initial C API Proposal
+
+This proposal is scoped to the C API bindings for the minimal 
+`ManagerInterface` methods that have been ported to C++ to date, 
+specifically `identifier()` and `displayName()`. Despite this tight
+restriction in scope there are several design decisions to be addressed.
+
+## Namespacing
+
+In C, exported symbols must be namespaced by prefixing the symbol name
+with some string.
+
+Since our top-level namespace includes a version tag, which is likely to
+change, we must use a macro to ease maintenance, e.g.
+
+```c++
+#define OPENASSETIO_NS(symbol) ...
+```
+such that `OPENASSETIO_NS(MyStruct)` expands to 
+`openassetio_v0_0_MyStruct`.
+
+## Strings
+
+Taking inspiration from [this conversation](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/194#discussion_r794047468)
+we can bundle a string buffer and length in a single C struct for error
+messages and parameters, e.g.
+
+```c++
+typedef struct {
+  const size_t maxSize;
+  char* buffer;
+  size_t usedSize;
+} OPENASSETIO_NS(SimpleString);
+```
+`usedSize` is for optimisation and safety and can technically be removed
+if we're happy to assume null-termination and accept the performance
+implications.
+
+## Opaque handles
+
+The opaque handle pattern used commonly in C APIs involves an arbitrary
+pointer to some user data, often coded simply as a `void*`. To add some
+meagre type safety, we can define an opaque handle as a pointer to an
+incomplete and ultimately unused class. We then `reinterpret_cast` it 
```

Comment:
You can get strong type safety with a wrapping pattern. i.e. put the type-erased pointer inside a strongly typed struct.

```
struct FooOpaque { void* p; };
```

Reply:
Ah yes, is there a compelling advantage of one approach over the other?  I guess `static_cast` can be used in your example, which is nice.

Reply:
Yes, there's a compelling advantage :)

```cpp
namespace clarity {
    // clarity in legibility:
    typedef void* EntityPtr;
    typedef void* ContainerPtr;

    // unsafe in that compiler doesn't care:
    ContainerPtr a = CreateContainer();
    EntityPtr b = CreateEntity();
    ContainerPtr c = a;   // compiler ignores error
    ContainerAddEntity(b, b); // compiler ignores error
}

namespace clarity_and_safety {
    struct EntityPtr { void* p };
    struct ContainerPtr { void* p };

    // safe in that compiler will disallow assignment:
    ContainerPtr a = CreateContainer();
    EntityPtr b = CreateEntity();
    ContainerPtr c = a;   // compiler reports error
    ContainerAddEntity(b, b); // compiler reports error
}
```
In both cases, the types are `sizeof(void*)` and can be passed with the same expense, and the pattern works for C and C++. The C++ side can have template and operator shenanigans to add utility to fetching out p such that the C++ code can look idiomatic without casting.

Reply:
Thanks for getting back to me. 

I guess by "compelling advantage" I meant versus the pointer-to-incomplete-type solution, i.e.
```c++
namespace clarity_and_safety_alternative {
    typedef struct UnusedIncompleteEntityType * EntityPtr;
    typedef struct UnusedIncompleteContainerType * ContainerPtr;

    ContainerPtr a = CreateContainer(); // Will `reinterpret_cast` real ptr to `ContainerPtr`
    EntityPtr b = CreateEntity(); // Will `reinterpret_cast` real ptr to `EntityPtr`
    ContainerPtr c = a;   // compiler reports error
    ContainerAddEntity(b, b); // compiler reports error
}
```

Reply:
Ah :) `struct Foo { void* p };` is handled as a type, whereas `typedef struct Foo* FooPtr` is handled as a pointer to a type. To clarify a bit, what I've been doing is `struct Foo { uintptr_t p; };` so that no magic in any ffi or other binding mechanism attempts to do ANYTHING with the pointer ~ it's completely type erased and binding languages tend to understand how to copy structs with POD guts.

Reply:
Thanks a lot for the clarification.  I really like the `uintptr_t` version, especially your argument about removing any ambiguity for FFI magic.  

We circulated the options internally, which of course just threw up yet more variations we hadn't considered :roll_eyes: 

But the TL;DR is that we're sticking with the classic opaque handle type, rather than wrapping a pointer in a struct, at least for now.

At least this is one of those decisions that doesn't require a major code overhaul if we change our mind.